// overwatch-api/lib/parser/profile.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function (platform, region, tag, callback) {
  _async2.default.auto({
    getHTML: _async2.default.apply(getHTML, platform, region, tag),
    parseHTML: ['getHTML', _async2.default.apply(parseHTML)],
    transform: ['getHTML', 'parseHTML', _async2.default.apply(transform)]
  }, function (err, results) {
    if (err) {
      return callback(err);
    }
    return callback(null, results.transform);
  });
};

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _async = require('async');

var _async2 = _interopRequireDefault(_async);

var _cheerio = require('cheerio');

var _cheerio2 = _interopRequireDefault(_cheerio);

var _request = require('request');

var _request2 = _interopRequireDefault(_request);

var _utils = require('./utils');

var _svg = require('./svg');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Get HTML from playoverwatch career page.
function getHTML(platform, region, tag, callback) {
  var url = platform === 'pc' ? 'https://playoverwatch.com/en-us/career/' + platform + '/' + region + '/' + tag : 'https://playoverwatch.com/en-us/career/' + platform + '/' + tag;

  var options = {
    uri: encodeURI(url),
    encoding: 'utf8'
  };

  return (0, _request2.default)(options, function (err, res, body) {
    return callback(err, body);
  });
}

// Begin html parsing.
function parseHTML(results, callback) {
  var $ = _cheerio2.default.load(results.getHTML);

  // Check if profile exists.
  var isFound = $('.content-box h1').text() !== 'Profile Not Found';
  if (!isFound) {
    return callback(new Error('Profile not found'));
  }

  var parsed = {
    user: $('.header-masthead').text(),
    level: $('.player-level div').first().text(),
    portrait: $('.player-portrait').attr('src'),
    permission: $('.masthead-permission-level-text').text(),
    endorsementLevel: $('.masthead .endorsement-level div').last().text(),
    endorsementFrame: $('.masthead .EndorsementIcon').attr('style').slice(21, -1),
    sportsmanshipValue: $('.masthead .EndorsementIcon-border--sportsmanship').data('value'),
    shotcallerValue: $('.masthead .EndorsementIcon-border--shotcaller').data('value'),
    teammateValue: $('.masthead .EndorsementIcon-border--teammate').data('value'),
    quickplayWonEl: $('#quickplay td:contains("Games Won")').next().html(),
    quickplayPlayedEl: $('#quickplay td:contains("Games Played")').next().html(),
    quickplayTimePlayedEl: $('#quickplay td:contains("Time Played")').next().html(),
    compWonEl: $('#competitive td:contains("Games Won")').next().html(),
    compPlayedEl: $('#competitive td:contains("Games Played")').next().html(),
    compLostEl: $('#competitive td:contains("Games Lost")').next().html(),
    compDrawEl: $('#competitive td:contains("Games Tied")').next().html(),
    compTimePlayedEl: $('#competitive td:contains("Time Played")').next().html(),
    compRankEl: $('.competitive-rank'),
    levelFrame: $('.player-level').attr('style').slice(21, -1),
    starEl: $('.player-rank').html(),
    rankEl: $('.player-level').html()
  };

  if (parsed.compRankEl !== null) {
    parsed.compRankImg = $('.competitive-rank img').attr('src') || null;
    parsed.compRank = $('.competitive-rank div').html();
  }

  if (parsed.starEl !== null) {
    parsed.star = $('.player-level .player-rank').attr('style').slice(21, -1);
  }

  if (parsed.rankEl !== null) {
    parsed.rank = $('.player-level').attr('style').slice(21, -1);
  }
  return callback(null, parsed);
}

// Transform the data into a json object we can serve.
function transform(results, callback) {
  var parsed = results.parseHTML;


  var endorsement = {
    sportsmanship: { value: parsed.sportsmanshipValue, rate: parseFloat((parsed.sportsmanshipValue * 100).toFixed(2)) },
    shotcaller: { value: parsed.shotcallerValue, rate: parseFloat((parsed.shotcallerValue * 100).toFixed(2)) },
    teammate: { value: parsed.teammateValue, rate: parseFloat((parsed.teammateValue * 100).toFixed(2)) },
    level: parseInt(parsed.endorsementLevel),
    frame: parsed.endorsementFrame
  };
  endorsement.icon = (0, _svg.createEndorsementSVG)(endorsement);

  var won = {};
  var lost = {};
  var draw = {};
  var played = {};
  var time = {};

  if (parsed.quickplayWonEl !== null) {
    won.quickplay = parsed.quickplayWonEl.trim().replace(/,/g, '');
  }

  if (parsed.quickplayPlayedEl !== null) {
    played.quickplay = parsed.quickplayPlayedEl.trim().replace(/,/g, '');
  }

  if (parsed.quickplayTimePlayedEl !== null) {
    time.quickplay = parsed.quickplayTimePlayedEl.trim().replace(/,/g, '');
  }

  if (parsed.compWonEl !== null) {
    won.competitive = parsed.compWonEl.trim().replace(/,/g, '');
  }

  if (parsed.compLostEl !== null) {
    lost.competitive = parsed.compLostEl.trim().replace(/,/g, '');
  }

  if (parsed.compDrawEl !== null) {
    draw.competitive = parsed.compDrawEl.trim().replace(/,/g, '');
  }

  if (parsed.compPlayedEl !== null) {
    played.competitive = parsed.compPlayedEl.trim().replace(/,/g, '');
  }

  if (parsed.compTimePlayedEl !== null) {
    time.competitive = parsed.compTimePlayedEl.trim().replace(/,/g, '');
  }

  // Calculate the prestige level.
  var level = parsed.level;
  if (parsed.star && parsed.rank) {
    var starsMatch = _path2.default.basename(parsed.star).split('.').slice(0, -1)[0];
    var rankMatch = _path2.default.basename(parsed.rank).split('.').slice(0, -1)[0];
    var stars = starsMatch ? (0, _utils.getPrestigeStars)(starsMatch) : 0;
    var rank = rankMatch ? (0, _utils.getPrestigeLevel)(rankMatch) : 0;
    var prestige = parseInt(stars) + parseInt(rank);
    level = parseInt(parsed.level) + parseInt(prestige) * 100;
  }

  var json = {
    username: parsed.user,
    level: parseInt(level),
    portrait: parsed.portrait,
    endorsement: endorsement,
    private: parsed.permission === 'Private Profile',
    games: {
      quickplay: {
        won: parseInt(won.quickplay),
        played: parseInt(played.quickplay) || undefined
      },
      competitive: {
        won: parseInt(won.competitive),
        lost: parseInt(lost.competitive),
        draw: parseInt(draw.competitive) || 0,
        played: parseInt(played.competitive),
        win_rate: parseFloat((parseInt(won.competitive) / parseInt(played.competitive - parseInt(draw.competitive)) * 100).toFixed(2))
      }
    },
    playtime: { quickplay: time.quickplay, competitive: time.competitive },
    competitive: { rank: parseInt(parsed.compRank), rank_img: parsed.compRankImg },
    levelFrame: parsed.levelFrame,
    star: parsed.star
  };
  return callback(null, json);
}
